File: mas-mcp/documentation-manager-server/python_conversion/src/documentation_manager/services/generator_service.py
from datetime import datetime
from typing import Dict, List, Optional, Any

from ..models.document import (
    Document,
    DocumentGenerationRequest,
    DocumentTemplate,
    TemplateSection,
)
from .document_service import DocumentService
from .template_service import TemplateService


class GeneratorService:
    """Service for document generation operations"""

    def __init__(self, template_service: TemplateService, document_service: DocumentService):
        self.template_service = template_service
        self.document_service = document_service

    async def generate_document(self, request: DocumentGenerationRequest) -> str:
        """Generate a document from a template and input data"""
        # Get template to use for generation
        template = None

        if request.template_id:
            template = await self.template_service.get_template(request.template_id)

        if not template:
            # Get default template for document type
            templates = await self.template_service.get_templates_by_type(request.type)
            if templates:
                template = templates[0]  # Use first available template
            else:
                raise ValueError(f"No template found for document type {request.type}")

        # Generate document content based on template and input data
        content = await self._generate_content_from_template(template, request.input_data)

        # Create document
        document = await self.document_service.create_document(request, content)

        return document.id

    async def _generate_content_from_template(
        self, template: DocumentTemplate, input_data: Dict[str, Any]
    ) -> str:
        """Generate document content based on template and input data"""
        # In a real implementation, this would use a language model to generate content
        # For now, we'll just create a simple markdown structure based on the template
        
        content = f"# {input_data.get('title', 'Untitled Document')}\n\n"

        # Add metadata
        content += "## Document Metadata\n\n"
        content += f"- Type: {template.type}\n"
        content += f"- Template: {template.name}\n"
        content += f"- Created: {datetime.utcnow().isoformat()}\n\n"

        # Process each section in the template
        for section in template.sections:
            content += self._generate_section(section, input_data, 2)

        return content

    def _generate_section(
        self, section: TemplateSection, input_data: Dict[str, Any], heading_level: int
    ) -> str:
        """Generate a section of content"""
        heading_marker = "#" * heading_level
        content = f"{heading_marker} {section.title}\n\n"

        # Check if we have specific content for this section in the input data
        section_key = section.title.lower().replace(" ", "_")

        if section_key in input_data:
            content += f"{input_data[section_key]}\n\n"
        else:
            # Add placeholder based on section description
            content += f"{section.description}\n\n"

        # Process subsections if any
        if section.subsections:
            for subsection in section.subsections:
                content += self._generate_section(subsection, input_data, heading_level + 1)

        return content

    async def generate_summary(self, document_id: str, max_length: int = 500) -> str:
        """Generate a summary of a document"""
        document = await self.document_service.get_document(document_id)
        if not document:
            raise ValueError(f"Document not found: {document_id}")

        # Simple summary implementation - just extract the first few characters
        # In a real implementation, this would use an LLM to generate a proper summary
        summary = document.content[:max_length]

        if len(document.content) > max_length:
            summary += "..."

        return summary 

File: mas-mcp/documentation-manager-server/src/services/generatorService.ts
import { DocumentGenerationRequest, DocumentTemplate, TemplateSection } from '../types.js';
import { TemplateService } from './templateService.js';
import { DocumentService } from './documentService.js';

export class GeneratorService {
  private templateService: TemplateService;
  private documentService: DocumentService;
  
  constructor(templateService: TemplateService, documentService: DocumentService) {
    this.templateService = templateService;
    this.documentService = documentService;
  }
  
  async generateDocument(request: DocumentGenerationRequest): Promise<string> {
    // Get template to use for generation
    let template: DocumentTemplate | null = null;
    
    if (request.templateId) {
      template = await this.templateService.getTemplate(request.templateId);
    } 
    
    if (!template) {
      // Get default template for document type
      const templates = await this.templateService.getTemplatesByType(request.type);
      if (templates.length > 0) {
        template = templates[0]; // Use first available template
      } else {
        throw new Error(`No template found for document type ${request.type}`);
      }
    }
    
    // Generate document content based on template and input data
    const content = await this.generateContentFromTemplate(template, request.inputData);
    
    // Save generated document
    const document = await this.documentService.createDocument(request, content);
    
    return document.id;
  }
  
  private async generateContentFromTemplate(template: DocumentTemplate, inputData: Record<string, any>): Promise<string> {
    // In a real implementation, this would use a language model to generate content
    // based on the template structure and input data
    
    // For now, we'll just create a simple markdown structure based on the template
    let content = `# ${inputData.title || 'Untitled Document'}\n\n`;
    
    // Add metadata
    content += `## Document Metadata\n\n`;
    content += `- Type: ${template.type}\n`;
    content += `- Template: ${template.name}\n`;
    content += `- Created: ${new Date().toISOString()}\n\n`;
    
    // Process each section in the template
    for (const section of template.sections) {
      content += this.generateSection(section, inputData, 2);
    }
    
    return content;
  }
  
  private generateSection(section: TemplateSection, inputData: Record<string, any>, headingLevel: number): string {
    const headingMarker = '#'.repeat(headingLevel);
    let content = `${headingMarker} ${section.title}\n\n`;
    
    // Check if we have specific content for this section in the input data
    const sectionKey = section.title.toLowerCase().replace(/\s+/g, '_');
    
    if (inputData[sectionKey]) {
      content += `${inputData[sectionKey]}\n\n`;
    } else {
      // Add placeholder based on section description
      content += `${section.description}\n\n`;
    }
    
    // Process subsections if any
    if (section.subsections && section.subsections.length > 0) {
      for (const subsection of section.subsections) {
        content += this.generateSection(subsection, inputData, headingLevel + 1);
      }
    }
    
    return content;
  }
  
  // This would be enhanced with an actual LLM-based generation in production
  async generateSummary(documentId: string, maxLength: number = 500): Promise<string> {
    const document = await this.documentService.getDocument(documentId);
    if (!document) {
      throw new Error(`Document not found: ${documentId}`);
    }
    
    // Simple summary implementation - just extract the first few characters
    // In a real implementation, this would use an LLM to generate a proper summary
    const summary = document.content.substring(0, maxLength);
    
    return summary + (document.content.length > maxLength ? '...' : '');
  }
} 